### ISSUE TITLE
[expo-background-fetch] App has stopped when is killed and background task try to execute

### ISSUE THREAD

**Issue by RodolfoGS** [expo/expo#10475]
### Summary of Issue 

I'm having the **_AppName_ has stopped** alert when use `expo-background-fetch` **AND** I have **Proguard** enabled.
The issue is similar but not is the same that was reported. All works great if Proguard is disabled, but when I enable Proguard the app crash in background.

This happen in Bare workflow, I'm not tested in Managed workflow. Only Android.

### Related Issues

Another related issues to `background-location`, but I have this with `expo-background-fetch` and **only** when **Proguard** is enabled.

- https://github.com/expo/expo/issues/9570
- https://github.com/expo/expo/issues/9288
- https://github.com/expo/expo/issues/8483

### Logcat

```
2020-09-30 07:53:43.632 5371-5371/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: xxx, PID: 5371
    java.lang.RuntimeException: java.lang.NullPointerException: Attempt to invoke interface method 'void l.b.a.b.a(android.content.Context, l.b.a.b$b, java.lang.Runnable, l.b.b.l.e)' on a null object reference
        at android.app.job.JobServiceEngine$JobHandler.handleMessage(JobServiceEngine.java:112)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'void l.b.a.b.a(android.content.Context, l.b.a.b$b, java.lang.Runnable, l.b.b.l.e)' on a null object reference
        at expo.modules.taskManager.h.a(Unknown Source:125)
        at expo.modules.taskManager.c.a(Unknown Source:2)
        at expo.modules.backgroundfetch.BackgroundFetchTaskConsumer.didExecuteJob(Unknown Source:8)
        at expo.modules.taskManager.h.b(Unknown Source:64)
        at expo.modules.taskManager.TaskJobService.onStartJob(Unknown Source:9)
        at android.app.job.JobService$1.onStartJob(JobService.java:71)
        at android.app.job.JobServiceEngine$JobHandler.handleMessage(JobServiceEngine.java:108)
```

### Environment - output of `expo diagnostics` & the platform(s) you're targeting

```
  Expo CLI 3.27.12 environment info:
    System:
      OS: macOS 10.15.5
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 14.4.0 - /usr/local/bin/node
      Yarn: 1.19.1 - /usr/local/bin/yarn
      npm: 6.14.5 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.9.3 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.0, DriverKit 19.0, macOS 10.15, tvOS 14.0, watchOS 7.0
      Android SDK:
        Android NDK: 21.3.6528147
    IDEs:
      Android Studio: 4.0 AI-193.6911.18.40.6514223
      Xcode: 12.0/12A7209 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~39.0.3 => 39.0.3
      react: 16.13.1 => 16.13.1
      react-native: ~0.63.3 => 0.63.3
    npmGlobalPackages:
      expo-cli: 3.27.12
    Expo Workflow: bare
```

```
expo-background-fetch: 8.5.0
expo-task-manager: 8.5.0
```

### Steps to Reproduce

1. Create a Bare project
2. Install `expo-background-fetch`
3. Create a task to run in background
4. Open Android Studio and enable Proguard (`def enableProguardInReleaseBuilds = true`)
5. Create a release sdk
6. Install in your device, open the app and kill the app.
7. Wait a few minutes to see the **_AppName_ has stopped**

### Expected Behavior vs Actual Behavior

**Expected Behavior**: background tasks works great
**Actual Behavior**: Error **AppName has stopped**

I tested with the following proguard rule but doesn't work.
```
-keep class expo.modules.taskManager { *; }
```

**Comment by byCedric** [comment]
Hi @RodolfoGS! Thanks for the report, unfortunately, I can't reproduce this exact error when I follow your steps. Could you share an example repo where this happens for you?

**Comment by RodolfoGS** [comment]
Hi @byCedric, I created a reproducible demo: https://github.com/RodolfoGS/expo-background-fetch-isseu10475

### How to reproduce.
1. Create a release APK from demo
2. Install APK on your device
3. **KILL THE APP** (too many important) and wait for 1 minute
4. You'll see the **TestBackgroundFetch has stopped** alert

---

### How I created this demo:
1. `expo init TestBackgroundFetch`
2. Select Bare workflow -> minimal
3. `cd TestBackgroundFetch`
4. `expo install expo-background-fetch`
5. `expo install expo-task-manager`
6. Add following code to `App.js`:
```
import * as TaskManager from 'expo-task-manager';
import * as BackgroundFetch from 'expo-background-fetch';

TaskManager.defineTask('demo-task', async () => {
  console.log('demo-task is running');
  return BackgroundFetch.Result.NewData;
});

BackgroundFetch.registerTaskAsync('demo-task', {
  minimumInterval: 60, // 1 min
  stopOnTerminate: false,
  startOnBoot: true,
})
  .then(() => alert('BackgroundFetch.registerTaskAsync success'))
  .catch(error => alert(`Error registerTaskAsync: ${error.message}`));
```
7. Enable proguard editing android/app/build.gradle: `def enableProguardInReleaseBuilds = true`
8. Add the following rule to `proguard-rules.pro` file to fix another bug that are described below
```
-keep class expo.modules.backgroundfetch.** { *; }
```
9. Create release apk (Android Studio: Build -> Generate Signed Bundle / APK...)
10. Install in your device
11. Open the app
12. **KILL THE APP** (too many important)
13. Wait 1 minute and you will see the **TestBackgroundFetch has stopped** alert

---
Additional to this, I found another bug when you run `BackgroundFetch.registerTaskAsync`.
```
Initializing task consumer 'e.a.a.c' failed. Inherited error: <init> [class android.content.Context, interface i.b.c.f.f]"
```

I fixed this bug adding this rule to proguard rules file https://github.com/RodolfoGS/expo-background-fetch-isseu10475/blob/master/android/app/proguard-rules.pro#L12
```
-keep class expo.modules.backgroundfetch.** { *; }
```

---
Tested on Android Nexus 6P - Android 8.1.0

**Comment by byCedric** [comment]
Awesome, yes, with this project I can successfully repro the issue! Thanks for the detailed write-up. @lukmccall it looks like this is also related to Proguard, is this also fixed by your last PR (#10421)?

**Comment by lukmccall** [comment]
@byCedric, I don't think that my PR fixed this bug. However, we now can add `DoNotStrip` annotation  to `BackgrondFetchTaskConsumer`... and this should fix the problem with `BackgroundFetch.registerTaskAsync`.

Unfortunately, it won't fix the problem with the `AppName has stopped` alert. 

Frankly speaking, I really don't understand why the `BackgrondFetchTaskConsumer` is dropped here 🤷🏼‍♂️ I'm not familiar with the task manager code.

**Comment by MrHazimAli** [comment]
hye @lukmccall , @byCedric 

is there any work around to resolve this issue at the moment? few things that I did wont solve this issue currently and Im not very familiar with native code side.

**Comment by byCedric** [comment]
Sorry for not updating this issue earlier, this should be fixed by PR #10410. If you want to see an example app with background location implemented on Android, [check out this](https://github.com/byCedric/office-marathon#readme).

Happy holidays!
